### PR TITLE
fix(page_sidebar): Improve page sidebar and mobile flow behavior

### PR DESCRIPTION
**Pull request by gadenbuie**
This PR fixes an issue with the height of the sidebar on mobile and improves the mobile flow behavior.

## Problem: Sidebar height on mobile

The interaction between the `height: auto` on the `.bslib-page-fill` container (applied to mobile-sized screens) and the `height: 100%` on the `.sidebar` container of the `page_sidebar()` layout caused unexpected behavior in both Safari and Chrome/Firefox.

In Safari, the sidebar would be the full height of the screen, but the children of the main container would resize depending on the state of the sidebar. In the video below, you can see that when the sidebar is expanded, it's `height: 100%` causes the sidebar layout container to resize to 100% of the viewport, causing the children of the main content area to resize as well. When the sidebar is collapsed and the `height: 100%` is removed, the elements return to their expected sizes. This causes jumpy plots around the transition.

https://github.com/rstudio/bslib/assets/5420529/074bc6d8-40c2-4850-95a2-c6edbf895cb4

This behavior results from Safari's unique interpretation of the behavior of a `height: 100%` element when its parent has `height: auto`.

Unfortunately, in Chrome and Firefox, we don't get the results we're expecting either. In the video below, in Firefox on mobile, you can see that the layout container uses the intrinsic height of the main contents, which are shorter than the full viewport height, and therefore the sidebar is shorter than expected.

https://github.com/rstudio/bslib/assets/5420529/862973b4-d821-4b55-ae23-a6dbedb9c25a

## Fix: Full height sidebar and mobile flow

The primary fix involves avoiding two things:

1. Don't set `height: 100%` on the `.sidebar` container.
1. Don't use `height: auto` in `page_fillable()` to toggle filling behavior

The `height: 100%` on the sidebar container is unnecessary. I think we may have needed it in a previous iteration and kept it out of caution.

The use of `height: auto` in page fillable is subtle, but probably not what we wanted for a more than just the behavior described above for a couple of reasons:

1. We want to have more control over where fillability is disabled. Currently, `page_fillable()` disables fillability on the outer container, but in the context of a sidebar we want to keep fillable behavior at the outer edge and instead disable fillability of the main content area of the sidebar layout.
2. Instead of setting `height: auto`, I considered of toggling `display: block` or `display: flex` but this presents another problem. If we move from a flex container to a block container, all of the `bslib-gap-spacing` classes will still be present and we'll lose the vertical spacing between elements.

In short, we need to toggle fill behavior while retaining `display: flex` and we'll need more than a single CSS property to control behavior.

In this PR, I added `.bslib-flow-mobile`, which generally speaking is a class that can be used anywhere to disable the fill behavior of its immediate children. This class is now applied to `page_fillable(fillable_mobile = FALSE)`. (Technically, it sets `flex: none;` on `.html-fill-item` direct children.)

Then, `.bslib-page-sidebar` received an exception to instead break fillability on the children of the page's sidebar main content area. Because the main content area can now scroll inside the page, we also need to adjust the collapse toggle position of that outer sidebar.

### Safari

After these changes, on Safari, the sidebar no longer causes the plot size to jump during transition.

https://github.com/rstudio/bslib/assets/5420529/cccf3c79-4912-4da2-acf1-e297b675902b

### Chrome/Firefox

On Chrome/Firefox, the sidebar is now full height, even if the elements inside are scrolling.

https://github.com/rstudio/bslib/assets/5420529/888363a8-ae1a-4315-8076-3dcff1d235c9

### Collapse toggle position

https://github.com/rstudio/bslib/assets/5420529/be9dc9d0-2101-483e-aa23-b36dcf687333

### Preview on real mobile

Here's a short preview on a real mobile phone (iPhone 13), with both `fillable_mobile = FALSE` and `fillable_mobile = TRUE`.

https://github.com/rstudio/bslib/assets/5420529/5bdf6342-d9ce-4ffa-a60b-89f06327cc63

## Flow utility classes

I think `.bslib-flow-mobile` is going to be generally useful, but I also see utility in having additional utility classes that can opt into flow behavior at various breakpoints. I added `.bslib-flow-sm`, `.bslib-flow-md` and `.bslib-flow-lg` that turn on flow behavior at or below the breakpoint suffix – i.e. `.bslib-flow-md` disables fillability for medium-sized screens below the `md` breakpoint.

### CODE REVIEW

File: R/page.R [hunk 1]
```diff
@@ -277,6 +277,7 @@ page_sidebar <- function(
       fillable = fillable,
       border = FALSE,
       border_radius = FALSE,
+      # tags$main(..., as_fill_carrier()),
```
Comment (gadenbuie) [inline R/page.R:280]: I'd love it if we have a `<main>` element, but for now...

```suggestion
```

File: inst/components/scss/sidebar.scss [hunk 2]
```diff
@@ -72,8 +74,6 @@ $bslib-sidebar-column-sidebar: Min(calc(100% - var(--_icon-size)), var(--_sideba
 
   > .sidebar {
     grid-column: 1 / 2;
-    width: 100%;
-    height: 100%;
     border-right: var(--_vert-border);
```
Comment (cpsievert) [inline inst/components/scss/sidebar.scss:77]: I think we should bring back the `width: 100%`

Comment (gadenbuie) [reply inst/components/scss/sidebar.scss:77]: I checked all of our rules and `.sidebar` always has both `grid-column` and `grid-row` properties, which drive the height and width of the sidebar container. I'm comfortable with both being redundant, but happy to leave `width` for now if you want.

File: inst/components/scss/sidebar.scss [hunk 3]
```diff
@@ -2,7 +2,7 @@ $bslib-sidebar-bg: rgba(var(--bs-emphasis-color-rgb, 0,0,0), 0.05) !default;
 $bslib-sidebar-fg: var(--bs-emphasis-color, black) !default;
 $bslib-sidebar-toggle-bg: rgba(var(--bs-emphasis-color-rgb, 0,0,0), 0.1) !default;
 $bslib-sidebar-border: var(--bs-card-border-width, #{$card-border-width}) solid var(--bs-card-border-color, #{$card-border-color}) !default;
-$bslib-sidebar-column-sidebar: Min(calc(100% - var(--_icon-size)), var(--_sidebar-width));
+$bslib-sidebar-column-sidebar: Min(calc(100% - var(--_padding-icon)), var(--_sidebar-width));
```
Comment (gadenbuie) [inline inst/components/scss/sidebar.scss:5]: In desktop mode, if the sidebar is too wide for the container, this ensures that the sidebar doesn't expand _too_ far, leaving one collapse icon's worth of space in the main area. Which is useful if there are nested sidebars. (Slightly tangential to this PR but I didn't want to forget to fix this.)

File: inst/components/scss/sidebar.scss [hunk 4]
```diff
@@ -65,15 +67,13 @@ $bslib-sidebar-column-sidebar: Min(calc(100% - var(--_icon-size)), var(--_sideba
     border-top-left-radius: 0;
     border-bottom-left-radius: 0;
     padding: var(--_padding);
-    transition: padding var(--_transition-easing-x) var(--_transition-duration);
+    transition: padding var(-- _transition-easing-x) var(--_transition-duration);
```
Comment (cpsievert) [inline inst/components/scss/sidebar.scss:70]: Was this change intentional?

Comment (gadenbuie) [reply inst/components/scss/sidebar.scss:70]: Weird! No, not sure how that snuck in there.

```suggestion
    transition: padding var(--_transition-easing-x) var(--_transition-duration);
```

File: inst/components/scss/page_fillable.scss [hunk 5]
```diff
@@ -10,7 +10,63 @@
...
+
+@include media-breakpoint-down(md) {
+  .bslib-flow-md > .html-fill-item {
+    flex: 0 0 auto;
+  }
+}
+
+@include media-breakpoint-down(lg) {
+  .bslib-flow-lg > .html-fill-item {
+    flex: 0 0 auto;
```
Comment (cpsievert) [inline inst/components/scss/page_fillable.scss:70]: Unless we have immediate plans to use and document these, I think I'd prefer to not commit to supporting them, especially right before a release. It kinda feels like we might want to revisit this concept but with height-related container query?

```suggestion
```

**Review by cpsievert — APPROVED** [review]

File: inst/components/scss/page_fillable.scss (hunk 5 above)
Comment (gadenbuie) [reply inst/components/scss/page_fillable.scss:70]: That makes sense. I think the `.bslib-flow-mobile` class is enough for this release and reasonably covers most of the intent here.
